### ISSUE TITLE
Wrong error from max-statements-per-line on version 2.10.1

### ISSUE THREAD

**Issue by onemen** [eslint/eslint#6173]
**What version of ESLint are you using?**
2.10.1

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**
```
module.exports = {
  "root": true,
  "env": {
    "browser": true,
    "es6": true
  },

  "parserOptions": {
    "ecmaVersion": 6
  },

  "rules": {
    "max-statements-per-line": [2, {"max": 1}],
  }
};
```

**What did you do? Please include the actual source code causing the issue.**
```
(function() {
  let bar = 1,
      foo = 2;
  if (bar) {
    if (foo > 1)
      foo--;
    else
      foo++;
  }
}());
```

**What did you expect to happen?**
No errors

**What actually happened? Please include the actual, raw output from ESLint.**
error on line 8: ```foo++;```
```
C:\test eslint\test1.js: line 8, col 7, Error - This line has too many statements. Maximum allowed is 1. (max-statements-per-line)

1 problem
```

**Comment by alberto** [comment]
This was introduced in #5993

**Comment by kaicataldo** [comment]
Was able to reproduce this locally. The rule actually completely ignores `else` statements if they are BlockStatements. Seems like a bug to me, especially considering the `if` is allowed.

**Comment by kaicataldo** [comment]
I can work on this.

**Comment by mysticatea** [comment]
I found a fun thing.

```js
/*eslint max-statements-per-line:[2, {max:1}]*/

if (foo > 1)
  foo--;
else
  foo++;    // error here
```

```js
/*eslint max-statements-per-line:[2, {max:1}]*/

if (foo > 1)
  foo--;
else
  foo++;    // no error!!

bar;
```

This is caused by traversal order.
In block body is traversed by `for` loop in BlockStatement handler. So it address inside of the `if` statement after `bar;`.

**Comment by bjb568** [comment]
Reproduced this. In my project I have an error for every time I use an else statement without brackets. Definitely broken.

**Comment by mysticatea** [comment]
I will work on this.
I think we need to rewrite the traversal way of this rule.

```js
/*eslint max-statements-per-line:[2, {max:2}]*/

a; if (b) { c; d; }
z;
```

Currently, this `d;` should be warned, but not.
Because the traversal order of this is `a;` → `if (b)` → `z;` → `c;` → `d;`.
